[20512] Topic Keys Documentation (#183)
eProsima/vulcanexus: docs/rst/tutorials/core/deployment/keys/topic_keys.rst

.. include:: ../../../../exports/alias.include

.. _topic_keys_tutorial:

Topic Keys Tutorial
===================

.. contents::
    :depth: 2
    :local:
    :backlinks: none

Background
----------

In Vulcanexus, a *Topic* is a communication channel used for publishing and subscribing to messages.
The value of data associated with a topic changes over time and each of these values are known as *data samples*.

> :ref:`topic_keys` refer to topics where each data sample is uniquely identified by a set of key fields (known as *data instance*).

Author: Yes, it is misexplained

> Unlike standard topics, where each data sample is distinct, keyed topics allow the user to manage and

Author: Technically, it is the right vision but I was just trying to explain it in simpler words. I will change it anyways

access data samples more efficiently by associating them with specific keys.
Please, refer to the documented section on :ref:`topic_keys` for a more detailed explanation.

Creating custom IDL messages
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> In ROS 2, users can define its own :ref:`custom messages<CustomInterfaces>`. In addition,

Reviewer: Please, only break lines after period, and do it always
```suggestion
In ROS 2, users can define their own :ref:`custom messages<CustomInterfaces>`.
In addition,
```

> a user can also create custom messages using the `IDL (Interface Definition Language) <https://www.omg.org/spec/IDL/4.2/About-IDL>`_

Reviewer: ```suggestion
it is also possible to create custom messages using the `IDL (Interface Definition Language) <https://www.omg.org/spec/IDL/4.2/About-IDL>`_
```

format following a standardized way of defining the message structure.
Next example depicts how to define a custom message using the IDL format in ROS 2:

> .. code-block:: bash

Reviewer: I'd out it's msg equivalent here so the modules thing is clear

    # ExampleMsg.idl
    module package_name {
      module msg {
        struct ExampleMsg {
          long field;
          short another_field;
          boolean yet_another_field;
        };
      };
    };

> As it can be seen, it is required to name the upper module with the same name as the package name containing it and
> with a nested module called *msg*. Within the *msg* module, user can define the message structure as desired using the IDL syntax
> and name it with the same name as the file.

Reviewer: ```suggestion
As it can be seen, using the IDL format requires naming the upper module with the same name as the package name containing it.
Furthermore, the message structure shall be named as the file containing it, and it shall be declared within an *msg* module nested in the package name one.
```

One of the advantages of defining messages in Interface Definition Language (IDL) is the ability to use annotations.
Annotations are metadata that provide additional information about IDL constructs such as modules,

> interfaces, operations, attributes, and data types and are used to convey details that are not part of the core

Reviewer: ```suggestion
interfaces, operations, attributes, and data types.
They are used to convey details that are not part of the core
```

> IDL syntax but are relevant for code generation, documentation, or other purposes.

Reviewer: They are part of the IDL syntax. I'd say they are metadata to the data structure definition

> Annotations in IDL typically follow the ``@`` symbol and can be applied to various IDL constructs.

Reviewer: At this point I think you can refer to the tutorial about communicating Vulcanexus and Fast DDS, where the way to generate type support from IDLs is shown. Also, you can hint here at the `@key` annotation to link this section with the next one

Creating Keyed Messages
^^^^^^^^^^^^^^^^^^^^^^^

> Keyed topics typically consist of one or more key fields within the data sample structure.

Reviewer: They do not consists of key fields per-se. Keyed topics exist when one or more fields in the data structure are annotated as keys

> These key fields serve as unique identifiers for each data sample published within the topic.

Reviewer: Again, not unique ID for data samples, but for topic instances instead

Author: I think the suggestion was applied. Now the text states `These key fields serve as unique identifiers for topic instances`

to organize and manage the data samples, facilitating efficient access, retrieval, and filtering
of data based on the specified key criteria.

The ``@key`` annotation designates a field as a key for the given topic type, which can have zero or more key fields
and can be applied to structure fields of various types:

* Any primitive, such as booleans, integers, characters, strings or sequences.
* Other existing or defined messages. If those have defined an inner key member, then those fields will we used as part of the key.
  Otherwise, the key will be the concatenation of all the fields.

In order to specify multiple keys, separate ``@key`` annotations are used.

> The following example shows how to define a keyed message using the IDL format in ROS 2:

Reviewer: Prefer Vulcanexus over ROS 2 in these cases
```suggestion
The following example shows how to define a keyed message using the IDL format in Vulcanexus:
```

.. code-block:: bash

    # KeyedMsgName.idl
    module package_name {
      module msg {
        struct KeyedMsgName {
          @key long key;
          string data;
        };
      };
    };

.. note::

    Currently, the only supported message format that can be annotated with ``@key`` is *.idl*.

>     Neither ``.msg`` nor ``.srv`` files support annotations.

Reviewer: ```suggestion
    Neither ``.msg`` nor ``.srv`` files support annotations yet.
```

Author: Good detail

In this tutorial, we will demonstrate the use of topic keys in Vulcanexus by creating a simple example in

> which we will simulate a system with two sensors (each one identified by a key), and a controller.

Reviewer: Weird sentence. Also, I think that the purpose of the tutorial should be the very first thing to be stated

Author: Will be left for the new introductory section

Prerequisites
-------------

* It is recommended to have a basic understanding of ``Topic Keys`` introduced in the :ref:`topic_keys` section.
* An up-to-date Vulcanexus installation using one of the following installation methods:

  * :ref:`linux_binary_installation`
  * :ref:`linux_source_installation`
  * :ref:`docker_installation`

Preparing the demo package
--------------------------

Lets start by setting up the Vulcanexus environment.
For this, there are two possible options:

#.  Running the Vulcanexus Docker image.

    Run the Vulcanexus Docker image with:

    .. code-block:: bash

>         xhost local:root
>         docker run \
>             -it \
>             ubuntu-vulcanexus:{DISTRO}-desktop

Reviewer: ```suggestion
        docker run -it --rm ubuntu-vulcanexus:{DISTRO}-desktop
```

    Then, within the container, source the Vulcanexus installation with:

    .. code-block:: bash

            source /opt/vulcanexus/{DISTRO}/setup.bash

#.  Running the tutorial on the local host. For this second option,

>     it is necessary to have installed the ``vucanexus-iron-desktop`` package,

Reviewer: I do not see the need of `desktop` here

    since this is the one that includes all the simulation tools, demos and tutorials.

    Source the following file to setup the Vulcanexus environment:

    .. code-block:: bash

            source /opt/vulcanexus/{DISTRO}/setup.bash

Retrieving the sources
^^^^^^^^^^^^^^^^^^^^^^

> 

Reviewer: Write a sentence here

    wget -O README.md https://raw.githubusercontent.com/eProsima/vulcanexus/{DISTRO}/docs/resources/tutorials/core/deployment/keys/README.md
    wget -O msg/SensorDataMsg.idl https://raw.githubusercontent.com/eProsima/vulcanexus/{DISTRO}/docs/resources/tutorials/core/deployment/keys/msg/SensorDataMsg.idl
    wget -O src/right_sensor.cpp https://raw.githubusercontent.com/eProsima/vulcanexus/{DISTRO}/docs/resources/tutorials/core/deployment/keys/src/right_sensor.cpp
    wget -O src/left_sensor.cpp https://raw.githubusercontent.com/eProsima/vulcanexus/{DISTRO}/docs/resources/tutorials/core/deployment/keys/src/left_sensor.cpp
    wget -O src/controller.cpp https://raw.githubusercontent.com/eProsima/vulcanexus/{DISTRO}/docs/resources/tutorials/core/deployment/keys/src/controller.cpp

The resulting directory structure should be:

.. code-block::

    ~/vulcanexus_ws
     ├──src
        ├── demo_keys_cpp
            ├── CMakeLists.txt
            ├── README.md
            ├── msg
            │   └── SensorDataMsg.idl
            ├── package.xml
            └── src
                ├── left_sensor.cpp
                ├── right_sensor.cpp
                └── controller.cpp

> Building the demo package

Reviewer: I'm missing a section analyzing the most important parts of the code, such as the IDL, the CMake macro used to generate the type, the fact that there is only one sub, etc. Furthermore, I think we need some more "complexity" in the sense that it is a bit unclear what I'm gaining here over say using the same type without keys for instance. I think we also need some type of conceptual drawing of the case

Reviewer: Not adressed, I thinks this is important

        .. code-block:: bash

            source ~/vulcanexus_ws/install/setup.bash
            ros2 run demo_keys_cpp controller

.. note::

    If a docker deployment was preferred, it would be necessary to attach the other two terminals to
    the running docker container before executing the above commands.
    This can be done by running ``docker exec -it <container_name> /bin/bash``.

The resulting output should be similar to the following:

.. raw:: html

    <video width=100% height=auto autoplay loop controls muted>
        <source src="../../../../../_static/resources/tutorials/core/deployment/keys/topic_keys_tutorial.mp4">
        Your browser does not support the video tag.
    </video>

One of the important differences to note against having used a standard topic is that, setting a reliable
communication and defining history QoS (Quality of Service) as ``KEEP_LAST 1``, if the controller late joins the application, it is
by no means guaranteed to receive the last data published by both of the sensors whereas by using topic keys
we assure that the last status of each instance (sensor) is received. This is because QoS are applied per data instance.

> This and further benefits can be explored in :ref:`benefits_of_topic_keys`.

Reviewer: We can showcase something with the history depth by having one pub publishing way faster than the other, and maybe showing that without keys a lot of samples from the slow sub are overridden

Author: Will be addressed in the example refactor

Background
----------

In Vulcanexus, a *Topic* is a communication channel used for publishing and subscribing to updates of object states.
The value of data associated with a topic changes over time and each of these values are known as *data samples*.

> :ref:`topic_keys` refer to topics where each data sample can update specific parts of the entire object state described by the topic (known as *instance*).

Reviewer: ```suggestion
:ref:`topic_keys` refer to topics where each data sample represent an update of the state of a specific object  (known as *instance*) among all those objects represented in the topic.
```

> Unlike standard topics, where each data sample updates the entire object state with every data sample, keyed topics allow the user to reduce the number of required resources (topics, along with its associated publisher and subscriber) by multiplexing into a single one.

Reviewer: ```suggestion
Unlike standard topics, where each data sample updates the entire object state with every data sample, keyed topics allow the user to reduce the number of required resources (topics, along with its associated publisher and subscriber) by multiplexing updates of several objects of the same kind into a single resource.
```

Please, refer to the documented section on :ref:`topic_keys` for a more detailed explanation.

//TODO: Add video

Creating custom IDL messages
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> In ROS 2, users can define their own :ref:`custom messages<CustomInterfaces>`.

Reviewer: ```suggestion
In Vulcanexus, users can define their own :ref:`custom messages<CustomInterfaces>`.
```

In addition, it is also possible to create custom messages using the `IDL (Interface Definition Language) <https://www.omg.org/spec/IDL/4.2/About-IDL>`_ format following a standardized way of defining the message structure.

> Next example depicts how to define a custom message using the IDL format in ROS 2:

Reviewer: ```suggestion
Next example depicts how to define a custom message using the IDL format in Vulcanexus:
```

> .. code-block:: bash

Reviewer: This calls for 2 tabs

    module package_name {
      module msg {
        struct ExampleMsg {
          string field;
          short another_field;
          double yet_another_field;
        };
      };
    };

    # ExampleMsg.msg equivalent
    string field;
    short another_field;
    double yet_another_field;

As it can be seen, using the IDL format requires naming the upper module with the same name as the package name containing it.

> Furthermore, the message structure shall be named as the file containing it, and it shall be declared within an *msg* module nested in the package name one.

Reviewer: ```suggestion
Furthermore, the message structure shall be named as the file containing it, and it shall be declared within an *msg* module nested in the package name one.
```
```suggestion
Furthermore, the message structure shall be named as the file containing it, and it shall be declared within a *msg* module nested in the package name one.
```

One of the advantages of defining messages in Interface Definition Language (IDL) is the ability to use annotations.
Annotations are metadata to the data structure definition that provide additional information about IDL constructs such as modules, interfaces, operations, attributes, and data types.
They are relevant for code generation, documentation, or other purposes.
Annotations in IDL typically follow the ``@`` symbol and can be applied to various IDL constructs.
Please, refer to the tutorial about :ref:`communicating Vulcanexus and Fast DDS <dds2vulcanexus_topic>`, where the way to generate types from IDL files is shown.

The ``@key`` annotation is used to designate a member as key, which is covered in the following section.

Creating Keyed Messages
^^^^^^^^^^^^^^^^^^^^^^^

Keyed topics exist when one or more fields in the data structure are annotated as keys.
These key fields serve as unique identifiers for topic instances.
to organize and manage the data samples, facilitating efficient access, retrieval, and filtering

> of data based on the specified key criteria.

Reviewer: Combine these two sentences

            source /opt/vulcanexus/{DISTRO}/setup.bash

#.  Running the tutorial on the local host. For this second option,

>     it is necessary to have installed the ``vucanexus-iron-base`` package,

Reviewer: ```suggestion
    it is necessary to have the ``vucanexus-iron-base`` package installed,
```

>     since this is the one that includes all the simulation tools, demos and tutorials.

Reviewer: ```suggestion
    it is necessary to have the ``vucanexus-iron-base`` package installed.
```

            ros2 launch demo_keys_cpp keyed_sensors_launch.py

    .. tab:: Shell 2 (Controller)

        .. code-block:: bash

            source ~/vulcanexus_ws/install/setup.bash
            #Wait until sensor[10] publishes the first data (10 secs)
            ros2 run demo_keys_cpp keyed_controller

The resulting output should be similar to the following:

.. raw:: html

    <video width=100% height=auto autoplay loop controls muted>
        <source src="../../../../../_static/resources/tutorials/core/deployment/keys/keyed_topic.mp4">
        Your browser does not support the video tag.
    </video>

In this final case, the controller is able to successfully recover the latest state of each sensor (data instance) when it joins the application.
In addition, it uses optimum resources (it only requires one topic and one subscription) and guarantees a minimum discovery overhead.
Hence, it is by using topic keys when we assure that the latest status of each instance (sensor) is received. This is because the Quality of Service settings are applied per data instance.
These and further benefits can be explored in :ref:`benefits_of_topic_keys`.

> 

Author: Uh, good one
